### PR TITLE
Add exceptions

### LINKED ISSUE #29: Implementing exceptions to be thrown by implementations
Here is my proposal for this : 

1. Introduce a `ExceptionInterface`, an empty Interface that must be implemented by all Exceptions of this package.
2. Introduce 2 abstracts classes `SerializationException` and `DeserializationException` that will extend `\RuntimeException` and implement `ExceptionInterface`.
3. Introduce `StatementSerializationException`, `StatementDeserializationException`, etc. classes that extends either `SerializationException` or `DeserializationException`. I think you understand where they should be thrown.

WDYT ?

### PR DESCRIPTION
**Pull request by Lctrs**
Close #29

### CODE REVIEW

File: src/Exception/ExtraAttributesException.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+namespace Xabbuh\XApi\Serializer\Exception;
+
+/**
+ * To be thrown when additional data is present that couldn't be interpreted.
+ *
+ * @author Jérôme Parmentier <jerome.parmentier@acensi.fr>
+ */
+class ExtraAttributesException extends DeserializationException
+{
+    public function __construct(array $missingAttributes, \Exception $previous = null)
```
Comment (xabbuh) [inline src/Exception/ExtraAttributesException.php]: `$missingAttributes` is not the best name in this context. :)

File: src/Exception/MissingRequiredAttributesException.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * MissingRequiredAttributesException.
+ *
+ * @author Jérôme Parmentier <jerome.parmentier@acensi.fr>
+ */
+class MissingRequiredAttributesException extends DeserializationException
+{
+    public function __construct(array $missingAttributes, \Exception $previous = null)
+    {
+        $msg = sprintf('Following missing attributes are required : "%s".', implode('", "', $missingAttributes));
```
Comment (xabbuh) [inline src/Exception/MissingRequiredAttributesException.php]: I would say "The following required attributes are missing: ".

File: src/Exception/StatementDeserializationException.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * (c) Christian Flothmann <christian.flothmann@xabbuh.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Xabbuh\XApi\Serializer\Exception;
+
+/**
+ * StatementDeserializationException.
```
Comment (xabbuh) [inline src/Exception/StatementDeserializationException.php]: Please remove comments like this from the class docblocks. :) They don't add any real value.

File: src/Exception/ExtraAttributesException.php (hunk 1 above)
Comment (Lctrs) [reply src/Exception/ExtraAttributesException.php]: The joy of copy/pasting :)

**Comment by xabbuh** [comment]
I just wonder when we would throw a `MissingAttributesException` or `ExtraAttributesException` and when one of the other specialised exceptions. Any idea on what to base this decision? It shouldn't be confusing to consumers of the libs which exceptions they have to account for.

**Comment by Lctrs** [comment]
Was thinking the same.

We could rely on specific exceptions thrown by the underneath serializer libraries. But I think it should be enough just to fill the `previous` attribute of the multiple `DeserializationException` and `SerializationException`.

**Comment by xabbuh** [comment]
What about dropping `MissingAttributesException` and `ExtraAttributesException` and use specific codes to indicate the cause of the exception (the base classes could defined `MISSING_ATTRIBUTE` and `EXTRA_ATTRIBUTE` constants to make them available in all concrete exceptions)?

**Comment by Lctrs** [comment]
What's the point ? Maybe we could just fill the `message` attribute of our exceptions with the one from the exceptions from the underneath libraries ?

**Comment by xabbuh** [comment]
I have no concrete use case yet. So we could even just drop the idea for now. Though I think it may at some point be useful to distinguish certain classes of errors that could occur during serialisation/deserialisation.

**Comment by Lctrs** [comment]
For now, our use case is just to respond with a bad request response if the serialization/deserialization fails. We can always come back to this if our needs evolve.

**Comment by xabbuh** [comment]
Fair enough, let's focus on the problems we need to solve now. :)

**Comment by xabbuh** [comment]
👍 Can you update the branch alias to be prepared for the 3.0 release? I will then merge and release tonight.

**Comment by xabbuh** [comment]
Oh well, I think I will also create some additional tests that enforce implementations to throw these new exceptions.

**Comment by Lctrs** [comment]
Updating the branch alias ? What do you mean ?

**Comment by xabbuh** [comment]
Sorry, I meant updating https://github.com/php-xapi/serializer/blob/master/composer.json#L36 to be `3.0.x-dev`.

**Comment by xabbuh** [comment]
Thank you @Lctrs.

**Pull request closed without merging**